### PR TITLE
🚧 Adopt the ‘to+infinitive’ Convention

### PR DESCRIPTION
**Pull request by jGleitz**
Let’s get the ball rolling!

This PR will introduce the [‘to+infinitive’ convention](https://github.com/robstoll/atrium-roadmap/issues/93) for atrium.
It will not be merged before the convention has been applied to all of the fluent API. 
It exists to gather feedback and find merge conflicts during the transition.

### CODE REVIEW

**Comment by jGleitz** [comment]
You’ll notice that I made cosmetical adaptions to the docs of the new functions. I think they are improvements and am happy to discuss them!

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:50]: Might needs to be in the text as it depends on the context whether this function throws or not. I am also happy with removing the `@throws AssertionError` from all functions since this adds little to the KDoc

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:77]: `is not the same`

I think we should either remove the `@throws` or keep the generic message. To cumbersome and error prone to maintain/keep in sync with the description IMO and there is no gain in repeating it, or is there?

**Review by robstoll — CHANGES_REQUESTED** [review]

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:47]: I am not entirely sure if `this` helps or confuses because by only reading this sentence, I would guess this function is an extension of `Assertion` due to the way `this` is wrapped with back-ticks. 
However, I think it would be an improvement if it is not wrapped with back-ticks and could be changed everywhere.

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:49]: Thanks for all the little improvements :+1: 
Same problematic as above with `this` and maybe it would also make sense to remove `current`? So my suggestion including your improvement:
`@return An [Expect] for the subject of this assertion.`

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:33]: Detail but more aligned with the readme:
```suggestion
 * Deprecated as Atrium moves to a consistent ‘to + infinitive’ naming convention. Use [toBeTheSameAs] instead.
```

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (jGleitz) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:50]: I think the doc either needs to state clearly when the exception is thrown, or not at all. ‘Might’ makes me mad because I’d like to know _when_.

Proposal:

> if the subject of `this` assertion is not the same instance as [expected] and this function is not called from inside an assertion creator lambda. If the call happens from inside an assertion creator lambda, the exception will be thrown by the function receiving the lambda.

But I am also completely fine with just removing all the `@throws`. I just think that if they are there, they need to be clear.

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (jGleitz) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:47]: I usually like writing ‘this’ as `this`, to point out that we are talking about the programming meaning of ‘this’. However, I agree that it is a little irritating because the receiver type is `Expect`.

Maybe this works:

 > Expects that the subject of `this` expectation is the same instance as [expected].

But I am also fine with just writing ‘this’, without the backticks. You’ll be the judge.

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (jGleitz) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:49]: I am very happy to leave out ‘current’.

I would like to keep the lowercase ‘an’ since what follows is a phrase started by the word ‘return’. To capitalize, it should be a proper sentence (see also the [Javadoc convention for @param](https://www.oracle.com/de/technical-resources/articles/java/javadoc-tool.html#@param)

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (jGleitz) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:77]: by the way, ‘is the same’ is correct here: The assertion throws if the instance are actually the same.

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:50]: So I suggest to remove them, I guess most people will expect that it throws an AssertionError and if in doubt, it is easy to find out if required. Less cluttering the code with something which has only little gain is a code thing IMO.

Btw. it is hard to come up with something which is actually true in all cases. 

> if the subject of this assertion is not the same instance as [expected] and this function is not called from inside an assertion creator lambda. If the call happens from inside an assertion creator lambda, the exception will be thrown by the function receiving the lambda. But only if it is not contributed to an ExplanatoryAssertionGroup or similar because such AssertionGroups usually do not cause an Exception... unless someone has explicitly marked them to fail. Yet, it still depends on the Expect if it is going to throw or not, in case of the default implementation, it is actually only the RootExpect which will throw... but you never know :laughing: 

And that's really not helpful either for a user -- I guess it is kind of helpful for a core contributor, to better understand when an exception is thrown but should probably be documented elsewhere.

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:47]: > Expects that the subject of `this` expectation is the same instance as [expected].

Sound good, we should replace `subject of the assertion` in all places with ``subject of  `this` expectation``

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:77]: true :slightly_smiling_face:

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:49]: So:
```
@return an [Expect] for the subject of `this` expectation.
```

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (jGleitz) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:47]: :+1:. I think we should try to adopt the lingo ‘expectation’ instead of ‘assertion’ everywhere.

Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:47]: Agreed. Let's wait with such changes until we merge into master to avoid conflicts. I create a project, where we can track such tasks

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:50]: I suggest we remove them also after the merge, created a task in the project:
https://github.com/robstoll/atrium/projects/1#column-12080551

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (jGleitz) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:47]: I’d suggest the opposite: Do these simple improvements now, because they are the smaller, easier changes, and start the renaming afterwards. What do you think?

Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:47]: Good as well if you create a PR targeting master, then we can do it before merging the renames into master. But we should wait until I merge the MapLike.contains as we need to change those files as well.

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:44]: @jGleitz IMO we should provide once again a script in order that users have a smooth migration experience. Do you see the draft release here? https://github.com/robstoll/atrium/releases/tag/untagged-c4eefbbdd8d22b72bc20

Comment (jGleitz) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:44]: > Do you see the draft release

No, I get a 404 for that link.

Writing a script feels like duplicating what the `@ReplaceWith` annotations do anyway. Isn’t there any IntelliJ functionality to say ‘perform all migrations’?

Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:44]: Not that I know of, you were able to choose `Replace for all` for a particular ReplaceWith but last time I wanted to use it I couldn't find the option. Moreover, from my experience, the applications of the ReplaceWith by Intellij are many times buggy and don't work as they should. I was always happy that I had the scripts. I'll put a draft of the release note into the wiki so that you could provide the replacements as well
Up to you if you want to help out or not. It's also fine if I collect/write down the replacements at the end of 0.16.0

Comment (jGleitz) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:44]: I see. I’ll prioritize work on my open pull requests, but will help on the script if I find the time.

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:41]: I plan to remove things maybe earlier than 1.0.0. They will certainly stay for a while but maybe we already remove it in 0.18.0 or 0.19.0
```suggestion
    "Replaced by toBeTheSameAs. Will be removed latest with version 1.0.0",
```

Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:41]: Done for existing `@Deprecated` annotations

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:47]: going to address this now.

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt
Comment (robstoll) [reply apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/anyAssertions.kt:49]: going to address this now.

**Comment by robstoll** [comment]
closing this in favour of #866 (other parts have been already merged into master via other PRs)

**Pull request closed without merging**
